core/src/main/java/com/datastax/oss/driver/internal/core/loadbalancing/helper/BasicLocalDcHelper.java

```diff
@@ -0,0 +1,115 @@
+import com.datastax.oss.driver.internal.core.context.InternalDriverContext;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BasicLocalDcHelper implements LocalDcHelper {
+
+  private static final Logger LOG = LoggerFactory.getLogger(BasicLocalDcHelper.class);
+
+  @NonNull protected final InternalDriverContext context;
+  @NonNull protected final DriverExecutionProfile profile;
+  @NonNull protected final String logPrefix;
+
+  public BasicLocalDcHelper(
+      @NonNull InternalDriverContext context,
+      @NonNull DriverExecutionProfile profile,
+      @NonNull String logPrefix) {
+    this.context = context;
+    this.profile = profile;
+    this.logPrefix = logPrefix;
+  }
+
+  /**
+   * This implementation fetches the local datacenter from the programmatic configuration API, or
+   * else, from the driver configuration. If no user-supplied datacenter can be retrieved, returns
+   * {@link Optional#empty empty}.
+   *
+   * @return The local datacenter from the configuration, or {@link Optional#empty empty} if none
+   *     found.
+   */
+  @Override
+  @NonNull
+  public Optional<String> discoverLocalDc() {
+    String localDc = context.getLocalDatacenter(profile.getName());
+    if (localDc != null) {
+      LOG.debug("[{}] Local DC set programmatically: {}", logPrefix, localDc);
+      checkLocalDatacenterCompatibility(localDc, context.getMetadataManager().getContactPoints());
+      return Optional.of(localDc);
+    } else if (profile.isDefined(DefaultDriverOption.LOAD_BALANCING_LOCAL_DATACENTER)) {
+      localDc = profile.getString(DefaultDriverOption.LOAD_BALANCING_LOCAL_DATACENTER);
+      LOG.debug("[{}] Local DC set from configuration: {}", logPrefix, localDc);
+      checkLocalDatacenterCompatibility(localDc, context.getMetadataManager().getContactPoints());
+      return Optional.of(localDc);
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  /**
+   * Checks if the contact points are compatible with the local datacenter specified either through
+   * configuration, or programmatically.
+   *
+   * <p>The default implementation logs a warning when a contact point reports a datacenter
+   * different from the local one.
+   *
+   * @param localDc The local datacenter, as specified in the config, or programmatically.
+   * @param contactPoints The contact points provided when creating the session.
+   */
+  protected void checkLocalDatacenterCompatibility(
+      @NonNull String localDc, Set<? extends Node> contactPoints) {
+    Set<Node> badContactPoints = new LinkedHashSet<>();
+    for (Node node : contactPoints) {
+      if (!Objects.equals(localDc, node.getDatacenter())) {
+        badContactPoints.add(node);
+      }
+    }
+    if (!badContactPoints.isEmpty()) {
```

Comment:
I think we should first validate that the local dc name is valid and in case it doesn't match, error out with a message like: `"Local data center 'my_dc' was provided but only found hosts in datacenters 'dc1' and 'dc2'"`

Reply:
`formatNodes()` already includes the DC name, nvm!